### ISSUE TITLE
[Enhancement] Save reading progress on Koreader Screensaver activation

### ISSUE THREAD

**Issue by mwozny123** [koreader/koreader#1990]
* KOReader version:
* Device:
Pocketbook, Kobo, probably all non-Kindle devices

#### Issue
For saving battery, non-Kindle devices use a auto power-off after specified time period.
 (Kindle has kind of hibernate).
 
When idle time is up - device system is shutdown, and Koreader is also terminated. 

Unfortunately if there was a reading progress, and Koreader was auto-terminated by system, it doesn't store reading progress.

#### Steps to reproduce

Turn on auto power-off, run Koreader, make some reading progress, wait until device is shutdown, start device, run Koreader. Koreader will not restore last reading progress, but the previously stored.

**Comment by houqp** [comment]
@koreader/kindle-dev , is there an event that we can get from `lipc-wait-event` for the power off action?

**Comment by NiLuJe** [comment]
@houqp: Which power off action? There isn't really one on Kindles, the ~7s thing goes to sleep directly and blanks the screen, IIRC, and more is a hard-reboot by cutting the power :?

**Comment by mwozny123** [comment]
This issue wasn't about kindle.

It was about pocketbook (and Kobo) where if you don't turn on power saving
(turn off device after certain amount of time, or each time turn it off
manually) in few days your battery will be empty, because the device will
be still running.

I don't remember how it is for Kobo but for Pocketbook there is also power
saving "sleep" option but if you use it without power saving "turn off"
option it doesn't help and next time (in couple days) you want to turn on
device - battery will be empty.

Work around for this is to always close koreader after finished reading,
but it's annoying, and many times when reading was suddenly interrupted
can't be done.

The request here is to implement kind of "on forced shutdown event", when
the current reading state, notes etc will be saved.
If this cannot be done maybe at least saving state after certain amount of
time or certain amount of read pages?
27 kwi 2016 2:39 AM "NiLuJe" <notifications@github.com> napisał(a):

**Comment by zlo2107** [comment]
Maybe the best solution is to add the save reading progress action to Koreader's screensaver (blank page). On Pocketbook devices ther is posibility to set "auto shut off". For example when i set this option to 5 minutes and the screen saver will start after e.g 2 minutes, there will be no problem with storing reading progress.

**Comment by houqp** [comment]
My bad, I misread the bug report. @mwozny123 can you confirm a screen saver always kicks in before the shutdown? If so, we can implement @zlo2107 's proposal.

I actually have a half done patch somewhere in my old laptop that implements write ahead commit log for book settings so we can recover from any crash. But I feel like this case should be handled by clean setting flush since it's not really a crash.

**Comment by mwozny123** [comment]
I think saving when screen saver is activated is a great idea.

But one question does Pocketbook have a Koreader Screensaver?
It for sure got one but I thought it was Pocketbook system Screensaver.

In Koreader "advanced settings" I found Kobo_screen_saver="".
Does it also apply to pocketbook?
Anyway I didn't find Screensaver timeout property.

I've a 10 min to turn off timeout. When it was reached Screensaver was
activated.
There was last read, with pocketbook system reader, book cover (different
than koreader last book).

My koreader version :
V2015.11-175-g7b11491

My bad, I misread the bug report. @mwozny123 <https://github.com/mwozny123>
can you confirm a screen saver always kicks in before the shutdown? If so,
we can implement @zlo2107 <https://github.com/zlo2107> 's proposal.

I actually have a half done patch somewhere in my old laptop that
implements write ahead commit log for book settings so we can recover from
any crash. But I feel like this case should be handled by clean setting
flush since it's not really a crash.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/koreader/koreader/issues/1990#issuecomment-215010500>

**Comment by houqp** [comment]
I don't have a pocketbook device, but according to the comment in our sleep handling code, we should be able to detect the device entered screen saver mode. Either way, I think this is a low hanging fruit feature that we should implement.

**Comment by mwozny123** [comment]
Would be great.
I'm just a little bit afraid, because system Screensaver(book cover) is
applied when system is going for shutdown.
It might be just a moment between changing to book cover a going for a
shutdown (while terminating all programs).
Will it still be ok?

If that would cause problem it could be also a good solution to apply
Koreader Screensaver (or only save reading progress) e.g. 1 minute before
system will go for a shutdown.
The time could be obtain from device settings (if possible) or entered
manually by user e.g. In Koreader "advanced settings".

IMHO something like "auto_save_after_inactive=9minutes" seems a better
solution that having 2 Screensavers. First one applied only after 9 minutes
for 1 minute (when system auto power off will start).
The time for battery saving (auto power off) is set up in device
configuration.
28 kwi 2016 09:09 "Qingping Hou" <notifications@github.com> napisał(a):

**Comment by houqp** [comment]
Writing settings to disk should take less than 1 second in most cases. I am trying to avoid adding more setting to the reader, especially for things that will rely on another system setting. Manually keeping them in sync is not a good experience.

**Comment by mwozny123** [comment]
Agree.
I would be happy to test that feature ;)

**Comment by mwozny123** [comment]
I'd changed the title to reflect topic arrangements.

**Comment by KenMaltby** [comment]
The title makes me think of the possibility of having the "Suspended" msg expanded to be "Suspended at Page xxx".  I assume that the saving on suspend would update the data used by the Book Status page, and that if it were displayed on suspend ...

**Comment by mwozny123** [comment]
About the issue title: @KenMaltby I'm not sure what exactly you are taking about. I
s the Screensaver just a message "Suspended"? 
Anyway could you change or (if you don't have permission)  write here a better title name? 
I'll change it.

**Comment by houqp** [comment]
Hm... actually, we are already doing book settings flush right before suspend: https://github.com/koreader/koreader/blob/master/frontend/device/generic/device.lua#L124. If wonder if it's because no event was send during auto shutoff.

**Comment by Frenzie** [comment]
I was just going to reply with that exact same line, lol.

**Comment by KenMaltby** [comment]
Please don't feel that a change in the title is being called for as a result of my musings, I was just given an idea by an alternative interpretation of the wording of the title.

Do the saved "book settings" include the data on "book status" (to me they sound like two different things)?
If there is an updated variable for current page, it seems like it could be added to the "show" for the "InfoMessage" at line 130 of screensaver.lua.  I was also thinking, if a static image of the Book status page were created, (like a screenshot) from the updated data, it could be added to the "self.suspend_msg" offerings.

**Comment by mwozny123** [comment]
I think the best solution would be if all of user "changes" in book were stored (so of course reading progress, created bookmarks, highlighted text etc). 

So after "battery-saving" power off - it can be restored exactly as it was before.

If it's too complex for one task maybe let's focus here at least on saving reading progress (last read page) and I'll create a new task for other bookmarks, hightlights etc?

Imho this is one of the most expected Pocketbook Koreader features.

**Comment by chrox** [comment]
The requested feature is already implemented, whenever the device goes into suspend koreader will flush settings to disk.
